### PR TITLE
Refactor resolver and fix download blob bug for localfs mode

### PR DESCRIPTION
**Pull request by zyfjeff**
1. use common resolver for localfs mode and startgz
2. fix download blobs don't auth for localfs mode

Signed-off-by: zyfjeff <zyfjeff@linux.alibaba-inc.com>

### CODE REVIEW

File: pkg/filesystem/fs/resolver.go [hunk 1]
```diff
@@ -48,21 +47,11 @@ func (r *Resolver) Resolve(ref, digest string, labels map[string]string) (io.Rea
...
-		tr = r.transport
-	} else {
-		tr, err = registry.AuthnTransport(nref, r.transport, keychain)
-		if err != nil {
-			return nil, errors.Wrapf(err, "failt to create authn transport %v", keychain)
-		}
+	url, tr, err := r.res.Resolve(nref, digest, keychain)
+
+	if err != nil {
+		return nil, errors.Wrapf(err, "failt to create authn transport %v", keychain)
```
Comment (imeoer) [inline pkg/filesystem/fs/resolver.go:53]: typo `failt` -> `failed`.

File: pkg/filesystem/fs/resolver.go [hunk 2]
```diff
@@ -12,24 +12,23 @@ import (
...
 	"github.com/containerd/nydus-snapshotter/pkg/auth"
-	"github.com/containerd/nydus-snapshotter/pkg/utils/registry"
+	"github.com/containerd/nydus-snapshotter/pkg/utils/transport"
 	"github.com/google/go-containerregistry/pkg/name"
 	retryablehttp "github.com/hashicorp/go-retryablehttp"
 	"github.com/pkg/errors"
 )
 
-const HTTPClientTimeOut = time.Second * 30
+const HTTPClientTimeOut = time.Second * 60
```
Comment (imeoer) [inline pkg/filesystem/fs/resolver.go:23]: Is this used?

Comment (zyfjeff) [reply pkg/filesystem/fs/resolver.go:23]: used, default HTTP retry client's retry timeout

Comment (zyfjeff) [reply pkg/filesystem/fs/resolver.go:23]:
```
func newRetryHTTPClient(tr http.RoundTripper) *retryablehttp.Client {
	retryClient := retryablehttp.NewClient()
	retryClient.HTTPClient.Transport = tr
	retryClient.HTTPClient.Timeout = HTTPClientTimeOut
	retryClient.Logger = nil
	return retryClient
}
```

File: pkg/filesystem/fs/resolver.go (hunk 1 above)
Comment (zyfjeff) [reply pkg/filesystem/fs/resolver.go]: fixed

File: pkg/utils/transport/pool.go [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+		ref.Context().RegistryStr(),
+		ref.Context().RepositoryStr(),
+		digest)
+
+	if tr, ok := r.trPool.Get(ref.Name()); ok {
+		if url, err := redirect(endpointURL, tr.(http.RoundTripper)); err != nil {
+			return url, tr.(http.RoundTripper), nil
+		}
+	}
+	r.trPool.Remove(ref.Name())
```
Comment (imeoer) [inline pkg/utils/transport/pool.go:55]: Remove a nonexistent key?

File: pkg/utils/transport/pool.go [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+	defer r.trPoolMu.Unlock()
+	endpointURL := fmt.Sprintf("%s://%s/v2/%s/blobs/%s",
+		ref.Context().Registry.Scheme(),
+		ref.Context().RegistryStr(),
+		ref.Context().RepositoryStr(),
+		digest)
+
+	if tr, ok := r.trPool.Get(ref.Name()); ok {
+		if url, err := redirect(endpointURL, tr.(http.RoundTripper)); err != nil {
+			return url, tr.(http.RoundTripper), nil
```
Comment (imeoer) [inline pkg/utils/transport/pool.go:52]: Maybe print a warning for the `err`.

File: pkg/utils/transport/pool.go [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+		io.Copy(ioutil.Discard, res.Body)
+		res.Body.Close()
+	}()
+
+	if res.StatusCode/100 == 2 {
+		url = endpointURL
+	} else if redir := res.Header.Get("Location"); redir != "" && res.StatusCode/100 == 3 {
+		url = redir
+	} else {
+		return "", fmt.Errorf("failed to access to %q with code %v", endpointURL, res.StatusCode)
```
Comment (imeoer) [inline pkg/utils/transport/pool.go:91]: In general, the response body includes some detailed error message, maybe we need to parse as text and print it out.

File: pkg/utils/transport/pool.go [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+
+	if tr, ok := r.trPool.Get(ref.Name()); ok {
+		if url, err := redirect(endpointURL, tr.(http.RoundTripper)); err != nil {
+			return url, tr.(http.RoundTripper), nil
+		}
+	}
+	r.trPool.Remove(ref.Name())
+	tr, err := registry.AuthnTransport(ref, r.transport, keychain)
+	if err != nil {
+		return "", nil, err
```
Comment (imeoer) [inline pkg/utils/transport/pool.go:58]: Wrap the error if necessary.

File: pkg/utils/transport/pool.go (hunk 3 above)
Comment (zyfjeff) [reply pkg/utils/transport/pool.go:55]: yep, remove invalid or non-existing key

**Comment by zyfjeff** [comment]
@imeoer ping

**Comment by zyfjeff** [comment]
Thanks for your review, I added some unittests for transport pool, PTAL @imeoer

File: pkg/utils/transport/pool.go [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+		io.Copy(ioutil.Discard, res.Body)
+		res.Body.Close()
+	}()
+
+	if res.StatusCode/100 == 2 {
+		url = endpointURL
+	} else if redir := res.Header.Get("Location"); redir != "" && res.StatusCode/100 == 3 {
+		url = redir
+	} else {
+		body, _ := io.ReadAll(res.Body)
```
Comment (imeoer) [inline pkg/utils/transport/pool.go:94]: Better to handle the error.

File: pkg/utils/transport/pool.go [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+		return "", errors.Wrapf(err, "failed to request to the registry")
+	}
+	req.Close = false
+	req.Header.Set("Range", "bytes=0-0")
+	res, err := tr.RoundTrip(req)
+	if err != nil {
+		return "", errors.Wrapf(err, "failed to request to %q", endpointURL)
+	}
+	defer func() {
+		io.Copy(ioutil.Discard, res.Body)
```
Comment (imeoer) [inline pkg/utils/transport/pool.go:85]: Better to give a warning for the error.

Comment (zyfjeff) [reply pkg/utils/transport/pool.go]: fixed

File: pkg/utils/transport/pool.go (hunk 7 above)
Comment (zyfjeff) [reply pkg/utils/transport/pool.go]: fixed

**Review by imeoer — APPROVED** [review]

**Pull request merged**